# Actividad 08: optimización y redes neuronales

### Sebastián Hernández Mantilla

## Optimización y redes neuronales

Se desarrolla el tutorial de reconocimiento de dígitos con el conjunto de datos MNIST de la Guia inicial de TensorFlow 2.0

In [187]:
import matplotlib.pyplot as plt
import tensorflow as tf
import numpy as np
import pandas as pd

mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 2s 869us/step - loss: 0.2970 - accuracy: 0.9136
Epoch 2/5
1875/1875 [==============================] - 2s 993us/step - loss: 0.1465 - accuracy: 0.9562
Epoch 3/5
1875/1875 [==============================] - 2s 805us/step - loss: 0.1105 - accuracy: 0.9667
Epoch 4/5
1875/1875 [==============================] - 2s 851us/step - loss: 0.0915 - accuracy: 0.9713
Epoch 5/5
1875/1875 [==============================] - 2s 993us/step - loss: 0.0762 - accuracy: 0.9768
313/313 - 0s - loss: 0.0791 - accuracy: 0.9742


[0.07908767461776733, 0.9742000102996826]

#### Función de pérdida

La función de pérdida utilizada en el tutorial es Categorical crossentropy:

![](imagenes/Captura.PNG)

Donde:

- N es el número de observaciones
- C es el número de categorias 
- Pmodel[yi ∈ C] es la probabilidad predecida por el modelo para que la observación i pertenezca a la categoría c


## Metodo de optimización
 Se analisa el desempeño de la red con diferentes optimizadores y variando la tasa de aprendizaje

![](imagenes/comparison.png)

Se encontró que 200 épocas son suficientes para hacer una comparación entre los tipos de optimizadores

#### SGD - 0.001


In [ ]:
optimizer= tf.keras.optimizers.SGD(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura1.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sgd0,001.png)

#### SGD - 0.1

In [ ]:
optimizer= tf.keras.optimizers.SGD(learning_rate=0.1)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura2.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sgd0,1.png)

#### SGD - 0.5

In [ ]:
optimizer= tf.keras.optimizers.SGD(learning_rate=0.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura3.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sgd0,5.png)

#### SGD - 0.9

In [ ]:
optimizer= tf.keras.optimizers.SGD(learning_rate=0.9)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura4.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sgd0,9.png)

#### SGD - 1.5

In [ ]:
optimizer= tf.keras.optimizers.SGD(learning_rate=1.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura5.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sgd1,5.png)

#### RSMprop - 0.001

In [ ]:
optimizer= tf.keras.optimizers.RMSprop(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura6.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/rsm0,001.png)

#### RSMprop - 0.1

In [ ]:
optimizer= tf.keras.optimizers.RMSprop(learning_rate=0.1)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura7.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/rsm0,1.png)

#### RSMprop - 0.5

In [ ]:
optimizer= tf.keras.optimizers.RMSprop(learning_rate=0.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura8.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/rsm0,5.png)

#### RSMprop - 0.9


In [306]:
optimizer= tf.keras.optimizers.RMSprop(learning_rate=0.9)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

Epoch 1/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.3797 - accuracy: 0.7897
Epoch 2/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.4620 - accuracy: 0.7704
Epoch 3/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.9441 - accuracy: 0.7691
Epoch 4/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.9845 - accuracy: 0.7473
Epoch 5/200
1875/1875 [==============================] - 3s 1ms/step - loss: 2.0554 - accuracy: 0.7536
Epoch 6/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.9489 - accuracy: 0.7321
Epoch 7/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.3544 - accuracy: 0.7269
Epoch 8/200
1875/1875 [==============================] - 2s 1ms/step - loss: 2.2296 - accuracy: 0.7175
Epoch 9/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.9907 - accuracy: 0.7244
Epoch 10/200
1875/1875 [==============================] - 2s 1ms/step - l

1875/1875 [==============================] - 3s 1ms/step - loss: 1.1385 - accuracy: 0.7408
Epoch 158/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.0170 - accuracy: 0.7423
Epoch 159/200
1875/1875 [==============================] - 2s 1ms/step - loss: 0.9866 - accuracy: 0.7419
Epoch 160/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.0467 - accuracy: 0.7416
Epoch 161/200
1875/1875 [==============================] - 3s 1ms/step - loss: 1.3893 - accuracy: 0.7498: 0s
Epoch 162/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.5140 - accuracy: 0.7438
Epoch 163/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.1800 - accuracy: 0.7399
Epoch 164/200
1875/1875 [==============================] - 2s 1ms/step - loss: 0.9174 - accuracy: 0.7429
Epoch 165/200
1875/1875 [==============================] - 2s 1ms/step - loss: 1.4443 - accuracy: 0.7517
Epoch 166/200
1875/1875 [==============================] - 2s 1ms

[15.115407943725586, 0.7864999771118164]

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sgd0,9.png)

#### RSMprop - 1.5

In [ ]:
optimizer= tf.keras.optimizers.RMSprop(learning_rate=1.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura9.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/rsm1,5.png)

#### Adagram - 0.001

In [ ]:
optimizer= tf.keras.optimizers.Adagrad(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura10.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/adagram0,001.png)

#### Adagram - 0.1

In [ ]:
optimizer= tf.keras.optimizers.Adagrad(learning_rate=0.1)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura11.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/adagram0,1.png)

#### Adagram - 0.5

In [ ]:
optimizer= tf.keras.optimizers.Adagrad(learning_rate=0.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura12.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/adagram_0,5.png)

#### Adagram - 0.9

In [ ]:
optimizer= tf.keras.optimizers.Adagrad(learning_rate=0.9)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura13.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/adagram0,9.png)

#### Adagram - 1.5


In [ ]:
optimizer= tf.keras.optimizers.Adagrad(learning_rate=1.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura14.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/adagram1,5.png)

#### AdDelta - 0.001

In [ ]:
optimizer= tf.keras.optimizers.Adadelta(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura15.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/adadeltal0,001.png)

#### AdDelta - 0.1

In [ ]:
optimizer= tf.keras.optimizers.Adadelta(learning_rate=0.1)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura16.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/adadelta0,1.png)

#### AdDelta - 0.5


In [ ]:
optimizer= tf.keras.optimizers.Adadelta(learning_rate=0.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura17.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/adadelta0,5.png)

#### AdDelta - 0.9


In [ ]:
optimizer= tf.keras.optimizers.Adadelta(learning_rate=0.9)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura18.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/adadelta0,9.png)

#### AdDelta - 1.5


In [ ]:
optimizer= tf.keras.optimizers.Adadelta(learning_rate=1.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train, y_train, epochs=200)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura19.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/adadelta1,5.png)

## Sobreentrenamiento
Se  reduce la cantidad de datos de entrenamiento y se reevalua el desempeño de los optimizadores. Se usó 600 datos para entrenamiento y 100 para testeo

#### SGD - 0.001

In [ ]:
optimizer= tf.keras.optimizers.SGD(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura20.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobresgd0,001.png)

#### SGD - 0.1

In [ ]:
optimizer= tf.keras.optimizers.SGD(learning_rate=0.1)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura21.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobresgd0,1.png)

#### SGD - 0.5

In [ ]:
optimizer= tf.keras.optimizers.SGD(learning_rate=0.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura22.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobresgd0,5.png)

#### SGD - 0.9

In [ ]:
optimizer= tf.keras.optimizers.SGD(learning_rate=0.9)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura23.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobresgd0,9.png)

#### SGD - 1.5

In [ ]:
optimizer= tf.keras.optimizers.SGD(learning_rate=1.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura24.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobresgd1,5.png)

#### RMSprop - 0.001

In [ ]:
optimizer= tf.keras.optimizers.RMSprop(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=100)
model.evaluate(x_test,  y_test, verbose=2)

![](imagenes/Captura25.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobrersm0,001.png)

#### RMSprop - 0.1

In [ ]:
optimizer= tf.keras.optimizers.RMSprop(learning_rate=0.1)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura26.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobrersm0,1.png)

#### RMSprop - 0.5

In [ ]:
optimizer= tf.keras.optimizers.RMSprop(learning_rate=0.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura27.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobrersm0,5.png)

#### RMSprop - 0.9

In [ ]:
optimizer= tf.keras.optimizers.RMSprop(learning_rate=0.9)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura28.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobrersm0,9.png)

#### RMSprop - 1.5

In [ ]:
optimizer= tf.keras.optimizers.RMSprop(learning_rate=1.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura29.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobrersm1,5.png)

#### Adagram - 0.001

In [ ]:
optimizer= tf.keras.optimizers.Adagrad(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura30.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobreadagram0,001.png)

#### Adagram - 0.1

In [ ]:
optimizer= tf.keras.optimizers.Adagrad(learning_rate=0.1)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura31.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobreadagram0,1.png)

#### Adagram - 0.5

In [ ]:
optimizer= tf.keras.optimizers.Adagrad(learning_rate=0.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura32.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobreadagram0,5.png)

#### Adagram - 0.9


In [ ]:
optimizer= tf.keras.optimizers.Adagrad(learning_rate=0.9)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura33.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobreadagram0,9.png)

#### Adagram - 1.5


In [ ]:
optimizer= tf.keras.optimizers.Adagrad(learning_rate=1.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura34.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobreadagram1,5.png)

#### AdDelta - 0.001

In [ ]:
optimizer= tf.keras.optimizers.Adadelta(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura35.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobreadadelta0,001.png)

#### AdDelta - 0.1

In [ ]:
optimizer= tf.keras.optimizers.Adadelta(learning_rate=0.1)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura36.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobreadadelta0,1.png)

#### AdDelta - 0.5

In [ ]:
optimizer= tf.keras.optimizers.Adadelta(learning_rate=0.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura37.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobreadadelta0,5.png)

#### AdDelta - 0.9

In [ ]:
optimizer= tf.keras.optimizers.Adadelta(learning_rate=0.9)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura38.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobreadadelta0,9.png)

#### AdDelta - 1.5

In [ ]:
optimizer= tf.keras.optimizers.Adadelta(learning_rate=1.5)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
#history = model.fit(X, Y, validation_split=0.33, epochs=150, batch_size=10, verbose=0)
history = model.fit(x_train_, y_train_, epochs=100)
model.evaluate(x_test_,  y_test_, verbose=2)

![](imagenes/Captura39.png)

In [ ]:
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 3))

#ax1.subplot(1, 2, 1)
ax1.plot(history.history['accuracy'])
#plt.plot(history.history['val_accuracy'])
ax1.set_title('model accuracy')
ax1.set_ylabel('accuracy')
ax1.set_xlabel('epoch')
#ax1.show()

#ax2.subplot(1, 2, 2)
ax2.plot(history.history['loss'])
#plt.plot(history.history['val_loss'])
ax2.set_title('model loss')
ax2.set_ylabel('loss')
ax2.set_xlabel('epoch')

![](imagenes/sobreadadelta1,5.png)

### Análisis

- ¿Cuál fue el método de optimización menos afectado por la reducción del conjunto de datos?
    
- ¿Cuál fue la tasa de aprendizaje menos afectada por la reducción del conjunto de datos?
    
- ¿Cuál fue el método que más rápido convergió?

In [199]:
df_sgd = pd.read_excel('comparacion.xlsx',sheet_name='Hoja2',index_col='Tasa aprendizaje')
df_rsm = pd.read_excel('comparacion.xlsx',sheet_name='Hoja3',index_col='Tasa aprendizaje')
df_adagram = pd.read_excel('comparacion.xlsx',sheet_name='Hoja4',index_col='Tasa aprendizaje')
df_adelta = pd.read_excel('comparacion.xlsx',sheet_name='Hoja5',index_col='Tasa aprendizaje')

df = pd.read_excel('comparacion.xlsx',sheet_name='Hoja6')

#### SGD

In [201]:
df_sgd

accuraccy 1  accuraccy 2
Tasa aprendizaje                          
0.001                   0.981        0.370
0.100                   0.981        0.370
0.500                   0.975        0.370
0.900                   0.968        0.370
1.500                   0.959        0.368

#### RSMprop

In [202]:
df_rsm

accuraccy 1  accuraccy 2
Tasa aprendizaje                          
0.001                   0.967        0.710
0.100                   0.947        0.627
0.500                   0.809        0.544
0.900                   0.786        0.505
1.500                   0.725        0.542

#### Adagram

In [203]:
df_adagram

accuraccy 1  accuraccy 2
Tasa aprendizaje                          
0.001                   0.724        0.542
0.100                   0.719        0.542
0.500                   0.724        0.542
0.900                   0.728        0.542
1.500                   0.725        0.586

#### AdDelta

In [204]:
df_adelta

accuraccy 1  accuraccy 2
Tasa aprendizaje                          
0.001                   0.725        0.588
0.100                   0.725        0.593
0.500                   0.730        0.598
0.900                   0.730        0.585
1.500                   0.730        0.560

Se promedió la perdida de precisión de cada método de optimización por las diferentes tasas de parendizaje y el optimizador que más precisión perdió con respecto al primer caso fue SGD pasando de un promedio de 0.97 a 0.36 
Para las tasas de aprendizaje se promedio el cambió de precisión para cada optimizador y se encontró que la que más se vio afectada por la reducción de datos fue la tasa de 0.1
El optimizador que más rápido converge, como se muestra en este análisis, es el RSMprop

## Bibliografía

- https://stats.stackexchange.com/questions/326065/cross-entropy-vs-sparse-cross-entropy-when-to-use-one-over-the-other
- https://github.com/keras-team/keras/issues/6444
- https://keras.io/api/optimizers/rmsprop/
- https://stats.stackexchange.com/questions/326065/cross-entropy-vs-sparse-cross-entropy-when-to-use-one-over-the-other
- https://keras.io/api/losses/
- https://machinelearningmastery.com/display-deep-learning-model-training-history-in-keras/
- https://towardsdatascience.com/neural-network-optimization-algorithms-1a44c282f61d